Bullish exchange integration

AresArtemius opened the pull request.

AresArtemius committed:
created bullish.ts

AresArtemius committed:
created abstract/bullish.ts

AresArtemius committed:
Merge branch 'ccxt:master' into bullish-exchange-integration

AresArtemius committed:
added fetchCurrencies

AresArtemius committed:
added fetchMarkets

AresArtemius committed:
Updated fetchMarkets

AresArtemius committed:
Updated fetchMarkets and added fetchOrderBook

AresArtemius committed:
Merge branch 'ccxt:master' into bullish-exchange-integration

AresArtemius committed:
added fetchTrades

AresArtemius committed:
added fetchTicker

AresArtemius committed:
added fetchOHLCV

AresArtemius committed:
added fetchOrders

AresArtemius committed:
added fetchOrder

AresArtemius committed:
added createOrder

AresArtemius committed:
added cancelOrder and cancelAllOrders

AresArtemius committed:
added fetchDepositsWithdrawals

AresArtemius committed:
updated sign and added signIn

AresArtemius committed:
fetchOrders fixed

AresArtemius committed:
added fetchAccounts

AresArtemius committed:
fixed fetchOrders

rayBastard committed:
req deleted

AresArtemius committed:
added fetchOrder

AresArtemius committed:
added fetchFundingRateHistory

AresArtemius committed:
fetchOrder and fetch Orders updated

AresArtemius committed:
fixed cancelAllOrders and flags

AresArtemius committed:
updated sign

AresArtemius committed:
updated cancelOrder

AresArtemius committed:
added withdraw

AresArtemius committed:
updated fetchTrades

AresArtemius committed:
added fetchMyTrades

AresArtemius committed:
added fetchBalance

AresArtemius committed:
Updated fetchBalance

AresArtemius committed:
added fetchPositions

rayBastard committed:
fetchPositionHistory added

rayBastard committed:
fetchTransfers added

rayBastard committed:
fetchBorrowRateHistory added

rayBastard committed:
add some comments

AresArtemius committed:
withdraw updated

AresArtemius committed:
parseMarket updated with expire

AresArtemius committed:
added Transfer

AresArtemius committed:
fixed FetchMarkets

AresArtemius committed:
fixed some mistake

AresArtemius committed:
fixed mistakes

AresArtemius committed:
fixed fetchTrades

AresArtemius committed:
currencies/bullish.json added

AresArtemius committed:
markets/bullish.json added

AresArtemius committed:
request/bullish.json added

AresArtemius committed:
fixed some mistakes

AresArtemius committed:
updated request-tests

AresArtemius committed:
added bullish.json

AresArtemius committed:
fixed some methods

AresArtemius committed:
updated request-tests

AresArtemius committed:
Updated response tests

AresArtemius committed:
updated request-tests

AresArtemius committed:
Updated fetchOHLCV

AresArtemius committed:
Updated parseTicker

AresArtemius committed:
Added pro/bullish.ts

AresArtemius committed:
Updated skipKeys in request-tests

AresArtemius committed:
Update parseTicker

AresArtemius committed:
added watchTicker

AresArtemius committed:
added watchOrderBook

AresArtemius committed:
updated fetchOrderBook

yzh-pelle committed:
handleOrderBook updated

yzh-pelle committed:
watchPrivate added

yzh-pelle committed:
fetchOHLCV updated with pagination

yzh-pelle committed:
fetchTrades updated with pagination

AresArtemius committed:
updated request-tests

AresArtemius committed:
updated fetchFundingRateHistory with pagination

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+        let symbol = undefined;
+        const baseId = this.safeString (market, 'baseSymbol');
+        const quoteId = this.safeString (market, 'quoteSymbol');
+        const base = this.safeString (market, 'baseSymbol');
```
base means the ccxt-unified base, we should convert it using safeCurrencyCode, we should not read it directly from the exchange market (the same applies to quote/settle/etc)

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+                'base': this.parseNumber (this.parsePrecision (basePrecision)),
+                'quote': this.parseNumber (this.parsePrecision (quotePrecision)),
+            },
+            'active': true,
```
if active is not available, we shouldn't populate it

AresArtemius replied:
done

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+    }
+
+    parseTicker (ticker: Dict, market: Market = undefined): Ticker {
+        const marketId = this.safeString (ticker, 'symbol');
```
same the example should be here

AresArtemius replied:
done

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+            return await this.fetchPaginatedCallDeterministic ('fetchOHLCV', symbol, since, limit, timeframe, params, maxLimit) as OHLCV[];
+        }
+        let until: Int = undefined;
+        [ until, params ] = this.handleOptionAndParams (params, 'fetchOHLCV', 'until');
```
we have the `handleUntilParam` method

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+        }
+        const stopPrice = this.safeString (params, 'stopPrice');
+        if (stopPrice !== undefined) {
+            request['stopPrice'] = stopPrice;
```
we can format the price here as well

AresArtemius replied:
done

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+            throw new ArgumentsRequired (this.id + ' cancelOrder() requires a symbol argument');
+        }
+        const market = this.market (symbol);
+        if (id === undefined) {
```
id is already a mandatory param, we don't need this

AresArtemius replied:
done

carlosmiei commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+     * @param {object} [params] extra parameters specific to the exchange API endpoint
+     * @returns response from exchange
+     */
+    async signIn (params = {}) {
```
@AresArtemius we shouldn't signIn every single time, we shoudl sign in once, cache the result and only sign in again after the current token expires (we also have to implement the refresh logic, if the current token is older than X minutes, signIn again"

AresArtemius replied:
done
delete signIn in methods

carlosmiei commented:
@AresArtemius I think we're missing the `handleErrors` implementation

carlosmiei commented on ts/src/pro/bullish.ts:
```diff
@@ -0,0 +1,347 @@
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const url = '/trading-api/v1/market-data/orderbook';
+        const messageHash = 'orderbook::' + symbol;
```
we shouldn't use the user-provided symbol for the messageHash because that might be the exchange-sepcific one, we should do `market['symbol']` or smth similar

carlosmiei commented on ts/src/pro/bullish.ts:
```diff
@@ -0,0 +1,347 @@
+            this.orderbooks[symbol] = this.orderBook ();
+        }
+        const orderbook = this.orderbooks[symbol];
+        const bids = this.separateBidsOrAsks (this.safeList (data, 'bids', []));
```
you sure we need this method? why?

yzh-pelle replied:
Both `price` and `amount` are in the same array without any key to distinguish.
Even indexes (including 0) are prices, odd indexes are amounts. I don't know how to handle it in a different way.

yzh-pelle committed:
update response-tests

AresArtemius committed:
updated src/bullish.ts

AresArtemius commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+        return this.parseMarkets (response);
+    }
+
+    parseMarket (market: Dict): Market {
```
done

AresArtemius commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+        } else {
+            throw new ArgumentsRequired (this.id + ' createOrder() requires a tradingAccountId parameter');
+        }
+        const stopPrice = this.safeString (params, 'stopPrice');
```
done

AresArtemius committed:
updated pro/bullish.ts

AresArtemius committed:
added handleErrors

yzh-pelle committed:
handleErrors updated

AresArtemius commented on ts/src/bullish.ts:
```diff
@@ -0,0 +1,2404 @@
+                'exact': {
+                },
+                'broad': {
+                },
```
Done

carlosmiei committed:
Revert "handleErrors updated"

This reverts commit ac5cf2cd14af8ca826d5f950f0020069b11a341b.

yzh-pelle committed:
handleErrors updated (correct file)

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into bullish-exchange-integration

carlosmiei committed:
python: read cookies from options

carlosmiei committed:
typescript/js cookies passing

yzh-pelle committed:
fetchFundingRateHistory deleted from request tests

https://github.com/ccxt/ccxt/actions/runs/16076838723/job/45373934923?pr=25884

yzh-pelle committed:
sign and signIn updated

yzh-pelle committed:
typo fixed in static response

yzh-pelle committed:
static request updated

yzh-pelle committed:
sign update reverted

due to https://github.com/ccxt/ccxt/actions/runs/16096718468/job/45420479251?pr=25884

yzh-pelle committed:
ping added

yzh-pelle committed:
watchOrders added

yzh-pelle committed:
parseOrder updated

yzh-pelle committed:
Response tests updated

yzh-pelle committed:
parseTrade updated

yzh-pelle committed:
fetchMyTrades and handleOrders added

yzh-pelle committed:
Response tests updated

yzh-pelle committed:
undefined option deleted

yzh-pelle committed:
watchPrivate, handleOrder and handleMyTrades updated

yzh-pelle committed:
watchBalance added

yzh-pelle committed:
parseMarket updated

yzh-pelle committed:
handlePong added

yzh-pelle committed:
Update bullish.ts

yzh-pelle committed:
handlePong updated

yzh-pelle committed:
parseMarket updated

yzh-pelle committed:
Update bullish.ts

yzh-pelle committed:
Update skip-tests.json with bullish exchange

yzh-pelle committed:
redundant import deleted

yzh-pelle committed:
parsePosition updated

yzh-pelle committed:
watchPositions added

yzh-pelle committed:
handleToken added

yzh-pelle committed:
watchPrivate updated

carlosmiei committed:
fix static tests

yzh-pelle committed:
methods with `since` and `until` rechecked and updated

carlosmiei committed:
change token type

carlosmiei committed:
php cookies working

carlosmiei committed:
c# cookie working

yzh-pelle committed:
handleMessage updated

yzh-pelle committed:
handleErrorMessage updated

yzh-pelle committed:
handleErrorMessage updated

yzh-pelle committed:
exceptions updated

yzh-pelle committed:
Merge branch 'master' into bullish-exchange-integration

yzh-pelle committed:
Exchagne.WsBridge.cs updated

Updated wsOptions initialization and removed cookie handling.

yzh-pelle committed:
ws/Clitnt.ts updated

yzh-pelle committed:
fix(bullish): cancelAllOrders signature updated

yzh-pelle committed:
feat(bullish): double `await` added to `Promise.all`

yzh-pelle committed:
feat(bullish): static markets updated

yzh-pelle committed:
feat(bullish): static currencies updated

yzh-pelle committed:
feat(bullish): createOrder updated

yzh-pelle committed:
feat(bullish): static markets updated

yzh-pelle committed:
feat(bullish): parseOrder and fetchDepositsWithdrawals updated

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
test(bullish): static response updated

Added trading account ID and new methods for creating and canceling orders.

yzh-pelle committed:
feat(bullish): fetchDepositAddress updated

yzh-pelle committed:
fix(bullish): fetchDepostAddress updated

yzh-pelle committed:
feat(bullish): features and some updates

yzh-pelle committed:
test(deepcoin): static request updated

Added fetchOrderTrades method to bullish.json

yzh-pelle committed:
test(bullish): static response updated

Added fetchOrderTrades method with sample response data.

yzh-pelle committed:
fix(bullish):  certified removed

yzh-pelle committed:
fix(bullish): duplicated id fixed

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
test(bullish): static response updated

yzh-pelle committed:
fix(deepcoin): skip-tests error fixed

yzh-pelle committed:
feat(bullish): parseMarket updated for option type of markets

yzh-pelle committed:
fix(bullish): js and php errors fixed

yzh-pelle committed:
fix(bullish): errors in parseMarket fixed

yzh-pelle committed:
test(bullish): static markets updated

yzh-pelle committed:
test(bullish): skip-tests updated

yzh-pelle committed:
feat(bullish): rate limits updated

yzh-pelle committed:
fix(bullish): revert changes after merging master branch

Return to #25884

yzh-pelle committed:
fix(bullish): remove headers from ws options

Removed empty headers object from options.

yzh-pelle committed:
feat(bullish): fetch orders by status added

yzh-pelle committed:
feat(bullish): new api endpoints added

yzh-pelle committed:
feat(bullish): abstract/bullish.ts updated

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into bullish-exchange-integration

carlosmiei committed:
fix token type

carlosmiei committed:
fix bullish ws go build

carlosmiei committed:
add missing types

yzh-pelle committed:
feat(bullish): editOrder added

yzh-pelle committed:
test(bullish): static request updated with editOrder

yzh-pelle committed:
test(bullish): static response updated with editOrder

yzh-pelle committed:
fix(bullish): request php error fixed

yzh-pelle committed:
fix(bullish): response php error fixed

yzh-pelle committed:
feat(deepcoin): methods for fetching orders updated

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
test(bullish): response tests updated

yzh-pelle committed:
fix(bullish): static request updated

yzh-pelle committed:
feat(bullish): fetchMyTrades updated

yzh-pelle committed:
feat(bullish): abstract/bullish.ts updated

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
test(bullish): static response updated

Added new methods for fetching order trades and my trades with detailed responses.

yzh-pelle committed:
fix(bullish): static response fixed

yzh-pelle committed:
fix(bullish): error in features fixed

yzh-pelle committed:
feat(bullish): pagination updated

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
fix(bullish): typo fixed

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
test(bullish): static response updated

yzh-pelle committed:
fix(bullish): some errors fixed

yzh-pelle committed:
fix(bullish): CS error fix

yzh-pelle committed:
feat(bullish): time difference added

yzh-pelle committed:
fix(bullish): remove changes

Changes of https://github.com/ccxt/ccxt/pull/25884/commits/9e5ab3ba131dc6ec99f6073f75bbc7ae0d5fe8e8 removed

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
feat(bullish): comments removed

yzh-pelle committed:
feat(bullish): fetchTransfers updated with pagination

yzh-pelle committed:
feat(bullish): remove privatePostV1CommandCommandTypeV1TransferAsset method

yzh-pelle committed:
test(bullish): static request updated

yzh-pelle committed:
test(bullish): static response updated

Added fetchTransfers method with detailed response structure.

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into bullish-exchange-integration

carlosmiei commented:
![bullish-logo-resized](https://github.com/user-attachments/assets/68f0686b-84f0-4da9-a751-f7089af3a9ed)

carlosmiei committed:
add logo

carlosmiei committed:
rm tradingAccount

carlosmiei committed:
update error message

carlosmiei committed:
add loadAccount
